http/src/v0/refs/path.rs

```diff
@@ -99,55 +99,73 @@ impl IpfsPath {
         self.root.take()
     }
 
+    pub fn resolve(&mut self, ipld: Ipld) -> Result<WalkSuccess, WalkFailed> {
+        let key = match self.next() {
+            Some(key) => key,
+            None => return Ok(WalkSuccess::EmptyPath(ipld)),
+        };
+
+        Self::resolve_segment(key, ipld)
+    }
+
+    pub fn resolve_segment(key: String, mut ipld: Ipld) -> Result<WalkSuccess, WalkFailed> {
+        ipld = match ipld {
+            Ipld::Link(cid) if key == "." => {
+                // go-ipfs: allows this to be skipped. lets require the dot for now.
+                // FIXME: this would require the iterator to be peekable in addition.
+                return Ok(WalkSuccess::Link(key, cid));
+            }
+            Ipld::Map(mut m) if m.contains_key(&key) => {
+                if let Some(ipld) = m.remove(&key) {
+                    ipld
+                } else {
+                    return Err(WalkFailed::UnmatchedMapProperty(m, key));
+                }
+            }
+            Ipld::List(mut l) => {
+                if let Ok(index) = key.parse::<usize>() {
+                    if index < l.len() {
+                        l.swap_remove(index)
+                    } else {
+                        return Err(WalkFailed::ListIndexOutOfRange(l, index));
+                    }
+                } else {
+                    return Err(WalkFailed::UnparseableListIndex(l, key));
+                }
+            }
+            x => return Err(WalkFailed::UnmatchableSegment(x, key)),
+        };
+
+        if let Ipld::Link(next_cid) = ipld {
+            Ok(WalkSuccess::Link(key, next_cid))
+        } else {
+            Ok(WalkSuccess::AtDestination(ipld))
+        }
+    }
+
```

Comment:
the two above are sort of "refactorings in preparation" as this should be going out of this complexified `impl Iterator<Item = String>` but I don't yet know how to put them or if they should be pushed upstream to rust-ipld.